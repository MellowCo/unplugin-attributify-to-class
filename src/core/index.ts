import { defaultRules, transformSelector } from 'unplugin-transform-class/utils'
import type { Options } from '../types'

const strippedPrefixes = [
  'v-bind:',
  ':',
  '@',
  'v-on',
]

const templateRe = /<template>([\s\S]*)<\/template>/g

const splitterRE = /[\s'"`;]+/g
const elementRE = /<\w(?=.*>)[\w:\.$-]*\s((?:".*?"|'.*?'|`.*?`|.*?)*?)>/gs
const valuedAttributeRE = /([?]|(?!\d|-{2}|-\d)[a-zA-Z0-9\u00A0-\uFFFF-\[\]#_:@.!%-]+)(?:=(["'])([^\2]*?)\2)?/g

const validateFilterRE = /(?!\d|-{2}|-\d)[a-zA-Z0-9\u00A0-\uFFFF-_:%-?]/

function isValidSelector(selector = ''): selector is string {
  return validateFilterRE.test(selector)
}

export const defaultAttributes = ['bg', 'flex', 'grid', 'border', 'text', 'font', 'class', 'className', 'p', 'm', 'animate', 'color', 'w', 'h', 'rd']
export const defaultIgnoreNonValuedAttributes = ['class', 'v-else']

interface TransformOption {
  /**
   * 元素标签
   */
  elementStr: string

  /**
   * 静态class
   */
  staticClass: string

  /**
   *  Attributify转class选择器
   */
  selectors: string[]
}

export function extractorAttributify(options?: Options): any {
  const attributes = options?.attributes ?? defaultAttributes
  const nonValuedAttribute = options?.nonValuedAttribute ?? true
  const ignoreNonValuedAttributes = options?.ignoreNonValuedAttributes ?? defaultIgnoreNonValuedAttributes
  const prefix = options?.prefix ?? 'un-'
  const prefixedOnly = options?.prefixedOnly ?? false
  const transformEscape = options?.transformEscape ?? true
  const transformRules = options?.transformRules ?? defaultRules
  const classPrefix = options?.classPrefix ?? ''

  return function extract(code: string) {
    const result: TransformOption[] = []

    // 只匹配 template 里的代码
    const templateMatchs = Array.from(code.matchAll(templateRe))

    if (templateMatchs.length) {
      const templateCode = templateMatchs[0][1]

      Array.from(templateCode.matchAll(elementRE))
        .forEach(([elementStr, valuedAttributeStr]) => {
          const valuedAttributes = Array.from((valuedAttributeStr || '').matchAll(valuedAttributeRE))

          const option: TransformOption = {
            elementStr,
            staticClass: '',
            selectors: [],
          }

          valuedAttributes.forEach(([sourceStr, name, _, content]) => {
            const _name = prefixedOnly ? name.replace(prefix, '') : name

            if (!content) {
              // fix: class="" ，生成 2 个 class 属性的问题
              if (name === 'class')
                option.staticClass = sourceStr

              // 是否只匹配 prefix
              if (prefixedOnly && !name.startsWith(prefix))
                return

              // 获取单值属性 <view p-2 m-2 />
              if (isValidSelector(_name) && nonValuedAttribute !== false) {
                // 是否为忽略的单值属性
                if (!ignoreNonValuedAttributes.includes(_name))
                  option.selectors.push(transformEscape ? transformSelector(`${classPrefix}${_name}`, transformRules) : `${classPrefix}${_name}`)
              }
              return
            }

            // : v-bind @ v-on 开头的属性 不处理
            for (const prefix of strippedPrefixes) {
              if (name.startsWith(prefix)) {
                name = name.slice(prefix.length)
                return
              }
            }

            // 是否生成该属性
            if (!attributes.includes(_name))
              return

            // 是否存在静态class
            if (['class', 'className'].includes(name)) {
              option.staticClass = sourceStr
            }
            else {
              // 是否只匹配 prefix
              if (prefixedOnly && !name.startsWith(prefix))
                return

              // 生成 class 选择器
              // bg="blue-400" => bg-blue-400
              const attributifyToClass = content
                .split(splitterRE)
                .filter(Boolean)
                .map((v) => {
                  if (v === '~')
                    return `${classPrefix}${_name}`

                  // support bg="active:red-400" => class="active:bg-red-400"
                  if (v.includes(':')) {
                    const [pseudoPrefix, pseudoValue] = v.split(':')
                    const classStr = `${pseudoPrefix}:${classPrefix}${_name}-${pseudoValue}`
                    return `${transformEscape ? transformSelector(classStr, transformRules) : classStr}`
                  }

                  return `${classPrefix}${_name}-${transformEscape ? transformSelector(v, transformRules) : v}`
                }).join(' ')

              option.selectors.push(attributifyToClass)
            }
          })

          result.push(option)
        })

      result.forEach(({ elementStr, staticClass, selectors }) => {
        if (selectors.length === 0)
          return

        if (staticClass) {
          const replaceStr = elementStr.replace(staticClass, spliceStr(staticClass, -1, ` ${selectors.join(' ')}`))
          code = code.replace(elementStr, replaceStr)
        }
        else {
          const classStr = ` class="${selectors.join(' ')}"`
          // fix: 自闭合标签 插入位置
          const insertIndex = elementStr.endsWith('/>') ? -2 : -1

          code = code.replace(elementStr, spliceStr(elementStr, insertIndex, classStr))
        }
      })
    }

    return code
  }
}

function spliceStr(str: string, start: number, newStr: string) {
  return str.slice(0, start) + newStr + str.slice(start)
}
