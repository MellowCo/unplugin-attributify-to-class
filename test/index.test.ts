import { describe, expect, test } from 'vitest'
import { defaultAttributes, defaultIgnoreNonValuedAttributes, extractorAttributify } from '../src/core'
import { IllegalStr, VElseStr, animateStr, classPrefixStr, emptyClassStr, escapeStr, fixture1, fixture2, fuiButton, noTemplateStr, placeholderStr, prefixAndClassPrefixStr, pseudoStr, scriptStr, valuelessStr } from './assets'

describe('attributify', () => {
  const attributesExtract = extractorAttributify({
    attributes: [
      ...defaultAttributes,
      'my-attr',
    ],
    nonValuedAttribute: false,
  })

  // 默认配置
  const defaultExtract = extractorAttributify()

  // 自定义前缀匹配
  const prefixExtract = extractorAttributify({
    prefixedOnly: true,
    prefix: 'un-',
    nonValuedAttribute: false,
  })

  // 忽略单值匹配
  const ignoreNonValueExtract = extractorAttributify({
    nonValuedAttribute: true,
    ignoreNonValuedAttributes: ['h-80', 'flex', 'my-prop'],
  })

  // 单值匹配
  const nonValueExtract = extractorAttributify({
    nonValuedAttribute: true,
  })

  // 完整配置
  const completeExtract = extractorAttributify({
    attributes: [...defaultAttributes, 'my-attr'],
    ignoreNonValuedAttributes: [...defaultIgnoreNonValuedAttributes, 'my-ignore'],
    nonValuedAttribute: true,
    prefix: 'li-',
    prefixedOnly: true,
  })

  // prefix and classPrefix
  const prefixAndClassPrefixExtract = extractorAttributify({
    prefix: 'pr-',
    prefixedOnly: true,
    classPrefix: 'cl-',
  })

  // 指定转换规则
  const customRulesextract = extractorAttributify({
    transformRules: {
      '.': '-dr-',
      '/': '-sr-',
      ':': '-cr-',
      '%': '-pr-',
      '!': '-er-',
      '#': '-wr-',
      '(': '-blr-',
      ')': '-brr-',
      '[': '-flr-',
      ']': '-frr-',
      '$': '-rr-',
    },
  })

  const classPrefixExtract = extractorAttributify({
    nonValuedAttribute: true,
    classPrefix: 'li-',
  })

  test('prefix and classPrefix', async () => {
    expect(defaultExtract(prefixAndClassPrefixStr)).toMatchSnapshot()

    expect(prefixAndClassPrefixExtract(prefixAndClassPrefixStr)).toMatchSnapshot()
  })

  test('attributesExtract', async () => {
    expect(attributesExtract(fixture1)).toMatchSnapshot()
    expect(attributesExtract(fixture2)).toMatchSnapshot()
  })

  test('prefixExtract', async () => {
    expect(prefixExtract(fixture1)).toMatchSnapshot()
    expect(prefixExtract(fixture2)).toMatchSnapshot()
  })

  test('ignoreNonValueExtract', async () => {
    expect(ignoreNonValueExtract(fixture1)).toMatchSnapshot()
    expect(ignoreNonValueExtract(fixture2)).toMatchSnapshot()
  })

  test('nonValueExtract', async () => {
    expect(nonValueExtract(fixture1)).toMatchSnapshot()
    expect(nonValueExtract(fixture2)).toMatchSnapshot()
    expect(nonValueExtract(animateStr)).toMatchSnapshot()
  })

  test('completeExtract', async () => {
    expect(completeExtract(fuiButton)).toMatchSnapshot()
  })

  test('escapeStr', () => {
    expect(attributesExtract(escapeStr)).toMatchSnapshot()
    expect(customRulesextract(escapeStr)).toMatchSnapshot()
  })

  test('scriptStr', () => {
    expect(nonValueExtract(scriptStr)).toMatchSnapshot()
    expect(nonValueExtract(noTemplateStr)).toMatchSnapshot()
  })

  test('emptyClassStr', () => {
    expect(nonValueExtract(emptyClassStr)).toMatchSnapshot()
  })

  test('IllegalStr', () => {
    expect(nonValueExtract(IllegalStr)).toMatchSnapshot()
  })

  test('classPrefixExtract', () => {
    expect(classPrefixExtract(classPrefixStr)).toMatchSnapshot()
  })

  test('valuelessStr', () => {
    expect(nonValueExtract(valuelessStr)).toMatchSnapshot()
  })

  test('pseudoStr', () => {
    expect(defaultExtract(pseudoStr)).toMatchSnapshot()
  })
  test('placeholderStr', () => {
    expect(defaultExtract(placeholderStr)).toMatchSnapshot()
  })

  test('VElseStr', () => {
    expect(defaultExtract(VElseStr)).toMatchSnapshot()
  })
})
