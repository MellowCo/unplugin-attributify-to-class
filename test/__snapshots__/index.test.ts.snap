// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`attributify > IllegalStr 1`] = `
"
<template>
  <input bg-green bg-red @input=\\"inputVal\\"   class=\\"bg-green bg-red\\"/>
  <input bg-green bg-red @input=\\"inputVal\\"   class=\\"bg-green bg-red\\">1111</>
</template>
"
`;

exports[`attributify > VElseStr 1`] = `
"
<template>
<view class='p-6' v-if='loading'>
</view>
<view class='p-6' v-else>
</view>
</template>
"
`;

exports[`attributify > attributesExtract 1`] = `
"
<template>
  <view >
    <button
      h-80 text-center flex flex-col align-center select-none all:transition-400
      bg=\\"blue-400 hover:blue-400 dark:!blue-400 dark:hover:blue-400\\"
      li-bg=\\"blue-500 hover:blue-500 dark:!blue-500 dark:hover:blue-500\\"
      un-bg=\\"red-500 hover:red-500 dark:!red-500 dark:hover:red-500\\"
      class=\\"text-red font-bold bg-blue-400 hover_cl_bg-blue-400 dark_cl_bg-_el_blue-400 dark_cl_bg-hover\\"
      :loading=\\"loading\\"
      my-class=\\"text-red font-bold\\"
      title=\\"this is title\\"
      my-prop
      non-value
      :class=\\"{ 'text-blue': true, 'text-green': false }\\"
      :class=\\"[ 'text-blue', 'text-green' ]\\"
      :class=\\"[ str === 'blue' ? 'text-blue' : 'text-red' ]\\"
    >
      Button
    </button>

    <button 
      text=\\"sm white\\"
      font=\\"mono light\\"
      p=\\"y-2 x-4\\"
      my-attr=\\"y-1 x-2 sm\\"
     class=\\"text-sm text-white font-mono font-light p-y-2 p-x-4 my-attr-y-1 my-attr-x-2 my-attr-sm\\">
      Button
    </button>

    <button 
      text=\\"sm white\\"
      font=\\"mono light\\"
      p=\\"y-2 x-4\\"
     class=\\"text-sm text-white font-mono font-light p-y-2 p-x-4\\">
      Button
    </button>

    <button border=\\"~ red\\" class=\\"border border-red\\">
      Button
    </button>

    <button flex=\\"~ col wrap\\" class=\\"flex flex-col flex-wrap\\">
      Button
    </button>

    <a text=\\"red\\" un-text=\\"blue\\" class=\\"text-red\\">This conflicts with links' text prop</a>

    <button 
      bg=\\"blue-400 hover:blue-500 dark:blue-500 dark:hover:blue-600\\"
      text=\\"sm white\\"
      font=\\"mono light\\"
      p=\\"y-2 x-4\\"
      border=\\"2 rounded blue-200\\"
     class=\\"bg-blue-400 hover_cl_bg-blue-500 dark_cl_bg-blue-500 dark_cl_bg-hover text-sm text-white font-mono font-light p-y-2 p-x-4 border-2 border-rounded border-blue-200\\">
      Button
    </button>

    <button 
      bg=\\"#333 grey-200 [#444]/40 red/50\\"
     class=\\"bg-_wn_333 bg-grey-200 bg-_lfl__wn_444_lfr__sl_40 bg-red_sl_50\\">
      Button
    </button>

    <image mode=\\"widthFix\\"></image> 
  </view>
</template>  
"
`;

exports[`attributify > attributesExtract 2`] = `
"
<template>
  <div h-80 text-center flex flex-col align-center select-none all:transition-400>
    <input type=\\"checkbox\\" peer mt-a>
    <div mb-a group peer-checked=\\"text-4xl\\">
      <div
        font-100 text-4xl mb--3 p-10
        bg-gradient=\\"to-r from-yellow-400 via-red-500 to-pink-500\\"
      >
        ~
      </div>
      <div text-5xl font-100 sm=\\"bg-blue-600\\">
        unocss
      </div>
      <div op-20 font-200 mt-1 tracking-wider group-hover=\\"text-teal-400 op-50\\">
        Re-imaging Atomic CSS
      </div>
    </div>
  </div>
  <div flex>
    <div ma inline-flex relative h-14>
      <input type=\\"text\\" m-0 pt-4 px-4 text-true-gray-800 peer 
      placeholder=\\"unocss\\" 
      un-placeholder=\\"text-red\\">
      <label absolute leading-1rem left-4 pointer-events-none text-gray-7 top=\\"1/3\\" transition=\\"200 linear\\"
        peer-not-placeholder-shown=\\"-translate-y-4 scale-75 origin-top-left text-green-500\\"
        peer-focus=\\"-translate-y-4 scale-75 origin-top-left text-green-500\\"
        before=\\"content-!\\"
        after=\\"content-[!]\\"
      >Experience now</label>
    </div>
  </div>
</template>

<script lang=\\"ts\\" setup>
import { onHide, onLaunch, onShow } from '@dcloudio/uni-app'
onLaunch(() => {
  console.log('App Launch')
})
onShow(() => {
  console.log('App Show')
})
onHide(() => {
  console.log('App Hide')
})
</script>

<style lang=\\"scss\\" scoped></style>
"
`;

exports[`attributify > classPrefixExtract 1`] = `
"
<template>
  <view bg-green bg-red text=\\"center left\\" class=\\"li-bg-green li-bg-red li-text-center li-text-left\\"></view>
</template>
"
`;

exports[`attributify > completeExtract 1`] = `
"
<template>
<view class=\\"fui-button__wrap\\" :class=\\"[!width || width==='100%' || width===true?'fui-button__flex-1':'']\\" :scope=\\"scope\\" @tap.stop=\\"handleTap\\" @click.stop=\\"handleTap\\" @click=\\"handleTap\\">
</view>
</template>
"
`;

exports[`attributify > emptyClassStr 1`] = `
"
<template>
  <view class=\\" bg-green bg-blue bg-red\\" bg-green bg=\\"blue red\\">
    <tm-skeleton-line :height=\\"props.height*4\\"></tm-skeleton-line>
  </view>
</template>
"
`;

exports[`attributify > escapeStr 1`] = `
"
<template>
  <button 
  bg=\\"#333 grey-200 [#444]/40 red/50\\" class=\\"bg-_wn_333 bg-grey-200 bg-_lfl__wn_444_lfr__sl_40 bg-red_sl_50\\">  
    Button
  </button>
</template>"
`;

exports[`attributify > escapeStr 2`] = `
"
<template>
  <button 
  bg=\\"#333 grey-200 [#444]/40 red/50\\" class=\\"bg--wr-333 bg-grey-200 bg--flr--wr-444-frr--sr-40 bg-red-sr-50\\">  
    Button
  </button>
</template>"
`;

exports[`attributify > ignoreNonValueExtract 1`] = `
"
<template>
  <view >
    <button
      h-80 text-center flex flex-col align-center select-none all:transition-400
      bg=\\"blue-400 hover:blue-400 dark:!blue-400 dark:hover:blue-400\\"
      li-bg=\\"blue-500 hover:blue-500 dark:!blue-500 dark:hover:blue-500\\"
      un-bg=\\"red-500 hover:red-500 dark:!red-500 dark:hover:red-500\\"
      class=\\"text-red font-bold text-center flex-col align-center select-none all_cl_transition-400 bg-blue-400 hover_cl_bg-blue-400 dark_cl_bg-_el_blue-400 dark_cl_bg-hover non-value\\"
      :loading=\\"loading\\"
      my-class=\\"text-red font-bold\\"
      title=\\"this is title\\"
      my-prop
      non-value
      :class=\\"{ 'text-blue': true, 'text-green': false }\\"
      :class=\\"[ 'text-blue', 'text-green' ]\\"
      :class=\\"[ str === 'blue' ? 'text-blue' : 'text-red' ]\\"
    >
      Button
    </button>

    <button 
      text=\\"sm white\\"
      font=\\"mono light\\"
      p=\\"y-2 x-4\\"
      my-attr=\\"y-1 x-2 sm\\"
     class=\\"text-sm text-white font-mono font-light p-y-2 p-x-4\\">
      Button
    </button>

    <button 
      text=\\"sm white\\"
      font=\\"mono light\\"
      p=\\"y-2 x-4\\"
     class=\\"text-sm text-white font-mono font-light p-y-2 p-x-4\\">
      Button
    </button>

    <button border=\\"~ red\\" class=\\"border border-red\\">
      Button
    </button>

    <button flex=\\"~ col wrap\\" class=\\"flex flex-col flex-wrap\\">
      Button
    </button>

    <a text=\\"red\\" un-text=\\"blue\\" class=\\"text-red\\">This conflicts with links' text prop</a>

    <button 
      bg=\\"blue-400 hover:blue-500 dark:blue-500 dark:hover:blue-600\\"
      text=\\"sm white\\"
      font=\\"mono light\\"
      p=\\"y-2 x-4\\"
      border=\\"2 rounded blue-200\\"
     class=\\"bg-blue-400 hover_cl_bg-blue-500 dark_cl_bg-blue-500 dark_cl_bg-hover text-sm text-white font-mono font-light p-y-2 p-x-4 border-2 border-rounded border-blue-200\\">
      Button
    </button>

    <button 
      bg=\\"#333 grey-200 [#444]/40 red/50\\"
     class=\\"bg-_wn_333 bg-grey-200 bg-_lfl__wn_444_lfr__sl_40 bg-red_sl_50\\">
      Button
    </button>

    <image mode=\\"widthFix\\"></image> 
  </view>
</template>  
"
`;

exports[`attributify > ignoreNonValueExtract 2`] = `
"
<template>
  <div h-80 text-center flex flex-col align-center select-none all:transition-400 class=\\"text-center flex-col align-center select-none all_cl_transition-400\\">
    <input type=\\"checkbox\\" peer mt-a class=\\"peer mt-a\\">
    <div mb-a group peer-checked=\\"text-4xl\\" class=\\"mb-a group\\">
      <div
        font-100 text-4xl mb--3 p-10
        bg-gradient=\\"to-r from-yellow-400 via-red-500 to-pink-500\\"
       class=\\"font-100 text-4xl mb--3 p-10\\">
        ~
      </div>
      <div text-5xl font-100 sm=\\"bg-blue-600\\" class=\\"text-5xl font-100\\">
        unocss
      </div>
      <div op-20 font-200 mt-1 tracking-wider group-hover=\\"text-teal-400 op-50\\" class=\\"op-20 font-200 mt-1 tracking-wider\\">
        Re-imaging Atomic CSS
      </div>
    </div>
  </div>
  <div flex>
    <div ma inline-flex relative h-14 class=\\"ma inline-flex relative h-14\\">
      <input type=\\"text\\" m-0 pt-4 px-4 text-true-gray-800 peer 
      placeholder=\\"unocss\\" 
      un-placeholder=\\"text-red\\" class=\\"m-0 pt-4 px-4 text-true-gray-800 peer\\">
      <label absolute leading-1rem left-4 pointer-events-none text-gray-7 top=\\"1/3\\" transition=\\"200 linear\\"
        peer-not-placeholder-shown=\\"-translate-y-4 scale-75 origin-top-left text-green-500\\"
        peer-focus=\\"-translate-y-4 scale-75 origin-top-left text-green-500\\"
        before=\\"content-!\\"
        after=\\"content-[!]\\"
       class=\\"absolute leading-1rem left-4 pointer-events-none text-gray-7\\">Experience now</label>
    </div>
  </div>
</template>

<script lang=\\"ts\\" setup>
import { onHide, onLaunch, onShow } from '@dcloudio/uni-app'
onLaunch(() => {
  console.log('App Launch')
})
onShow(() => {
  console.log('App Show')
})
onHide(() => {
  console.log('App Hide')
})
</script>

<style lang=\\"scss\\" scoped></style>
"
`;

exports[`attributify > nonValueExtract 1`] = `
"
<template>
  <view >
    <button
      h-80 text-center flex flex-col align-center select-none all:transition-400
      bg=\\"blue-400 hover:blue-400 dark:!blue-400 dark:hover:blue-400\\"
      li-bg=\\"blue-500 hover:blue-500 dark:!blue-500 dark:hover:blue-500\\"
      un-bg=\\"red-500 hover:red-500 dark:!red-500 dark:hover:red-500\\"
      class=\\"text-red font-bold h-80 text-center flex flex-col align-center select-none all_cl_transition-400 bg-blue-400 hover_cl_bg-blue-400 dark_cl_bg-_el_blue-400 dark_cl_bg-hover my-prop non-value\\"
      :loading=\\"loading\\"
      my-class=\\"text-red font-bold\\"
      title=\\"this is title\\"
      my-prop
      non-value
      :class=\\"{ 'text-blue': true, 'text-green': false }\\"
      :class=\\"[ 'text-blue', 'text-green' ]\\"
      :class=\\"[ str === 'blue' ? 'text-blue' : 'text-red' ]\\"
    >
      Button
    </button>

    <button 
      text=\\"sm white\\"
      font=\\"mono light\\"
      p=\\"y-2 x-4\\"
      my-attr=\\"y-1 x-2 sm\\"
     class=\\"text-sm text-white font-mono font-light p-y-2 p-x-4\\">
      Button
    </button>

    <button 
      text=\\"sm white\\"
      font=\\"mono light\\"
      p=\\"y-2 x-4\\"
     class=\\"text-sm text-white font-mono font-light p-y-2 p-x-4\\">
      Button
    </button>

    <button border=\\"~ red\\" class=\\"border border-red\\">
      Button
    </button>

    <button flex=\\"~ col wrap\\" class=\\"flex flex-col flex-wrap\\">
      Button
    </button>

    <a text=\\"red\\" un-text=\\"blue\\" class=\\"text-red\\">This conflicts with links' text prop</a>

    <button 
      bg=\\"blue-400 hover:blue-500 dark:blue-500 dark:hover:blue-600\\"
      text=\\"sm white\\"
      font=\\"mono light\\"
      p=\\"y-2 x-4\\"
      border=\\"2 rounded blue-200\\"
     class=\\"bg-blue-400 hover_cl_bg-blue-500 dark_cl_bg-blue-500 dark_cl_bg-hover text-sm text-white font-mono font-light p-y-2 p-x-4 border-2 border-rounded border-blue-200\\">
      Button
    </button>

    <button 
      bg=\\"#333 grey-200 [#444]/40 red/50\\"
     class=\\"bg-_wn_333 bg-grey-200 bg-_lfl__wn_444_lfr__sl_40 bg-red_sl_50\\">
      Button
    </button>

    <image mode=\\"widthFix\\"></image> 
  </view>
</template>  
"
`;

exports[`attributify > nonValueExtract 2`] = `
"
<template>
  <div h-80 text-center flex flex-col align-center select-none all:transition-400 class=\\"h-80 text-center flex flex-col align-center select-none all_cl_transition-400\\">
    <input type=\\"checkbox\\" peer mt-a class=\\"peer mt-a\\">
    <div mb-a group peer-checked=\\"text-4xl\\" class=\\"mb-a group\\">
      <div
        font-100 text-4xl mb--3 p-10
        bg-gradient=\\"to-r from-yellow-400 via-red-500 to-pink-500\\"
       class=\\"font-100 text-4xl mb--3 p-10\\">
        ~
      </div>
      <div text-5xl font-100 sm=\\"bg-blue-600\\" class=\\"text-5xl font-100\\">
        unocss
      </div>
      <div op-20 font-200 mt-1 tracking-wider group-hover=\\"text-teal-400 op-50\\" class=\\"op-20 font-200 mt-1 tracking-wider\\">
        Re-imaging Atomic CSS
      </div>
    </div>
  </div>
  <div flex class=\\"flex\\">
    <div ma inline-flex relative h-14 class=\\"ma inline-flex relative h-14\\">
      <input type=\\"text\\" m-0 pt-4 px-4 text-true-gray-800 peer 
      placeholder=\\"unocss\\" 
      un-placeholder=\\"text-red\\" class=\\"m-0 pt-4 px-4 text-true-gray-800 peer\\">
      <label absolute leading-1rem left-4 pointer-events-none text-gray-7 top=\\"1/3\\" transition=\\"200 linear\\"
        peer-not-placeholder-shown=\\"-translate-y-4 scale-75 origin-top-left text-green-500\\"
        peer-focus=\\"-translate-y-4 scale-75 origin-top-left text-green-500\\"
        before=\\"content-!\\"
        after=\\"content-[!]\\"
       class=\\"absolute leading-1rem left-4 pointer-events-none text-gray-7\\">Experience now</label>
    </div>
  </div>
</template>

<script lang=\\"ts\\" setup>
import { onHide, onLaunch, onShow } from '@dcloudio/uni-app'
onLaunch(() => {
  console.log('App Launch')
})
onShow(() => {
  console.log('App Show')
})
onHide(() => {
  console.log('App Hide')
})
</script>

<style lang=\\"scss\\" scoped></style>
"
`;

exports[`attributify > nonValueExtract 3`] = `
"
<template>
  <view animate-shake-y animate-iteration-infinite animate-duration-1500 class=\\"animate-shake-y animate-iteration-infinite animate-duration-1500\\"></view>
  <view animate=\\"shake-x iteration-infinite duration-1500\\" class=\\"animate-shake-x animate-iteration-infinite animate-duration-1500\\"></view>
</template>
"
`;

exports[`attributify > placeholderStr 1`] = `
"
<template>
<input type=\\"text\\"  placeholder=\\"What's your name?\\" />

<view
  text=\\"sm green\\"
  p=\\"y-2 x-4\\"
  m=\\"4\\"
 class=\\"text-sm text-green p-y-2 p-x-4 m-4\\">
xx
</view>


<view flex=\\"~ col wrap\\" class=\\"m4 flex flex-col flex-wrap\\">
  xxx
</view>




</template>
"
`;

exports[`attributify > prefix and classPrefix 1`] = `
"
<template>
  <view 
    text=\\"sm white\\"
    p=\\"y-4 x-4\\"
    bg-blue
    mx-4
    pr-text=\\"xl black\\"
    pr-p=\\"y-8 x-8\\"
    pr-bg-red
    pr-mx-8
   class=\\"text-sm text-white p-y-4 p-x-4 bg-blue mx-4 pr-bg-red pr-mx-8\\">
    Button  
  </view>
</template>
"
`;

exports[`attributify > prefix and classPrefix 2`] = `
"
<template>
  <view 
    text=\\"sm white\\"
    p=\\"y-4 x-4\\"
    bg-blue
    mx-4
    pr-text=\\"xl black\\"
    pr-p=\\"y-8 x-8\\"
    pr-bg-red
    pr-mx-8
   class=\\"cl-text-xl cl-text-black cl-p-y-8 cl-p-x-8 cl-bg-red cl-mx-8\\">
    Button  
  </view>
</template>
"
`;

exports[`attributify > prefixExtract 1`] = `
"
<template>
  <view >
    <button
      h-80 text-center flex flex-col align-center select-none all:transition-400
      bg=\\"blue-400 hover:blue-400 dark:!blue-400 dark:hover:blue-400\\"
      li-bg=\\"blue-500 hover:blue-500 dark:!blue-500 dark:hover:blue-500\\"
      un-bg=\\"red-500 hover:red-500 dark:!red-500 dark:hover:red-500\\"
      class=\\"text-red font-bold bg-red-500 hover_cl_bg-red-500 dark_cl_bg-_el_red-500 dark_cl_bg-hover\\"
      :loading=\\"loading\\"
      my-class=\\"text-red font-bold\\"
      title=\\"this is title\\"
      my-prop
      non-value
      :class=\\"{ 'text-blue': true, 'text-green': false }\\"
      :class=\\"[ 'text-blue', 'text-green' ]\\"
      :class=\\"[ str === 'blue' ? 'text-blue' : 'text-red' ]\\"
    >
      Button
    </button>

    <button 
      text=\\"sm white\\"
      font=\\"mono light\\"
      p=\\"y-2 x-4\\"
      my-attr=\\"y-1 x-2 sm\\"
    >
      Button
    </button>

    <button 
      text=\\"sm white\\"
      font=\\"mono light\\"
      p=\\"y-2 x-4\\"
    >
      Button
    </button>

    <button border=\\"~ red\\">
      Button
    </button>

    <button flex=\\"~ col wrap\\">
      Button
    </button>

    <a text=\\"red\\" un-text=\\"blue\\" class=\\"text-blue\\">This conflicts with links' text prop</a>

    <button 
      bg=\\"blue-400 hover:blue-500 dark:blue-500 dark:hover:blue-600\\"
      text=\\"sm white\\"
      font=\\"mono light\\"
      p=\\"y-2 x-4\\"
      border=\\"2 rounded blue-200\\"
    >
      Button
    </button>

    <button 
      bg=\\"#333 grey-200 [#444]/40 red/50\\"
    >
      Button
    </button>

    <image mode=\\"widthFix\\"></image> 
  </view>
</template>  
"
`;

exports[`attributify > prefixExtract 2`] = `
"
<template>
  <div h-80 text-center flex flex-col align-center select-none all:transition-400>
    <input type=\\"checkbox\\" peer mt-a>
    <div mb-a group peer-checked=\\"text-4xl\\">
      <div
        font-100 text-4xl mb--3 p-10
        bg-gradient=\\"to-r from-yellow-400 via-red-500 to-pink-500\\"
      >
        ~
      </div>
      <div text-5xl font-100 sm=\\"bg-blue-600\\">
        unocss
      </div>
      <div op-20 font-200 mt-1 tracking-wider group-hover=\\"text-teal-400 op-50\\">
        Re-imaging Atomic CSS
      </div>
    </div>
  </div>
  <div flex>
    <div ma inline-flex relative h-14>
      <input type=\\"text\\" m-0 pt-4 px-4 text-true-gray-800 peer 
      placeholder=\\"unocss\\" 
      un-placeholder=\\"text-red\\">
      <label absolute leading-1rem left-4 pointer-events-none text-gray-7 top=\\"1/3\\" transition=\\"200 linear\\"
        peer-not-placeholder-shown=\\"-translate-y-4 scale-75 origin-top-left text-green-500\\"
        peer-focus=\\"-translate-y-4 scale-75 origin-top-left text-green-500\\"
        before=\\"content-!\\"
        after=\\"content-[!]\\"
      >Experience now</label>
    </div>
  </div>
</template>

<script lang=\\"ts\\" setup>
import { onHide, onLaunch, onShow } from '@dcloudio/uni-app'
onLaunch(() => {
  console.log('App Launch')
})
onShow(() => {
  console.log('App Show')
})
onHide(() => {
  console.log('App Hide')
})
</script>

<style lang=\\"scss\\" scoped></style>
"
`;

exports[`attributify > pseudoStr 1`] = `
"
<template>
<view bg=\\"red-200 active:red-400\\" class=\\"bg-red-200 active_cl_bg-red-400\\"/>
</template>
"
`;

exports[`attributify > scriptStr 1`] = `
"
<template>
  <view bg=\\"#333\\" p=\\"x-6 y-10\\" w100 h200 class=\\"bg-_wn_333 p-x-6 p-y-10 w100 h200\\">this is a unocss</view>
</template>

<script lang=\\"ts\\" setup>
const CheckboxGropup = ref<InstanceType<typeof tmCheckboxGroup> | null>(null)
</script>
"
`;

exports[`attributify > scriptStr 2`] = `
"
<script lang=\\"ts\\" setup>
const CheckboxGropup = ref<InstanceType<typeof tmCheckboxGroup> | null>(null)
const a = 1
</script>
"
`;

exports[`attributify > valuelessStr 1`] = `
"
<template>
  <view class=\\"flex-1 bg-blue text-_lfl_16px_lfr_ text-center text-100_pes_ text-1_dl_2\\" text-[16px] text-center text-100% text-1.2>text</view>
  <view class=\\"flex-1 bg-_lfl__wn_ccc_lfr_\\" bg-[#ccc]>text</view>
</template>
"
`;
